vllm/model_executor/guided_decoding/__init__.py

```diff
@@ -1,25 +1,47 @@
-from typing import Optional, Union
+import asyncio
+import concurrent.futures
+from typing import Optional
 
-from vllm.entrypoints.openai.protocol import (ChatCompletionRequest,
-                                              CompletionRequest)
-from vllm.model_executor.guided_decoding.lm_format_enforcer_decoding import (
-    get_lm_format_enforcer_guided_decoding_logits_processor)
+from vllm.model_executor.guided_decoding.fields import GuidedDecodingFields
 from vllm.model_executor.guided_decoding.outlines_decoding import (
     get_outlines_guided_decoding_logits_processor)
 from vllm.sampling_params import LogitsProcessor
 
+global_thread_pool = None
 
-async def get_guided_decoding_logits_processor(
-        guided_decoding_backend: str, request: Union[CompletionRequest,
-                                                     ChatCompletionRequest],
-        tokenizer) -> Optional[LogitsProcessor]:
-    if guided_decoding_backend == 'outlines':
-        return await get_outlines_guided_decoding_logits_processor(
+
+async def get_guided_decoding_logits_processor_async(
+        request: GuidedDecodingFields, tokenizer) -> Optional[LogitsProcessor]:
+    global global_thread_pool
+    if global_thread_pool is None:
+        global_thread_pool = concurrent.futures.ThreadPoolExecutor(
+            max_workers=4)
+    loop = asyncio.get_running_loop()
+
+    return await loop.run_in_executor(
+        global_thread_pool,
+        get_guided_decoding_logits_processor,
+        request,
+        tokenizer,
+    )
+
+
+def get_guided_decoding_logits_processor(
```

Comment:
For the offline use case it makes sense to have a non-async way to get the processor. But orthogonal to that is the issue of sharing of stateful logits processors. In PR #5329 both concerns are addressed because it adds a factory with an async and a sync builder method.